ld_preload_shim.c

```diff
@@ -0,0 +1,427 @@
+    const rmw_publisher_options_t *publisher_options);
+
+typedef rmw_ret_t (*rmw_destroy_publisher_t)(rmw_node_t *node,
+                                             rmw_publisher_t *publisher_ptr);
+
+typedef rmw_ret_t (*rmw_publish_t)(const rmw_publisher_t *pub_ptr,
+                                   const void *message,
+                                   rmw_publisher_allocation_t *allocation);
+
+typedef rmw_ret_t (*clock_gettime_t)(clockid_t clockid, struct timespec *tp);
+
+typedef rmw_subscription_t *(*rmw_create_subscription_t)(
+    const rmw_node_t *node_ptr,
+    const rosidl_message_type_support_t *type_support, const char *topic_name,
+    const rmw_qos_profile_t *qos_policies,
+    const rmw_subscription_options_t *subscription_options);
+
+typedef rmw_ret_t (*rmw_destroy_subscription_t)(const rmw_node_t *node_ptr,
+                                                rmw_subscription_t *sub_ptr);
+
+typedef rmw_ret_t (*rmw_take_with_info_t)(
+    const rmw_subscription_t *subscription, void *ros_message, bool *taken,
+    rmw_message_info_t *message_info,
+    rmw_subscription_allocation_t *allocation);
+
+typedef rmw_service_t *(*rmw_create_service_t)(
+    const rmw_node_t *node, const rosidl_service_type_support_t *type_support,
+    const char *service_name, const rmw_qos_profile_t *qos_profile);
+
+typedef rmw_ret_t (*rmw_destroy_service_t)(rmw_node_t *node,
+                                           rmw_service_t *service);
+
+typedef rmw_client_t *(*rmw_create_client_t)(
+    const rmw_node_t *node, const rosidl_service_type_support_t *type_support,
+    const char *service_name, const rmw_qos_profile_t *qos_policies);
+
+typedef rmw_ret_t (*rmw_destroy_client_t)(rmw_node_t *node,
+                                          rmw_client_t *client);
+
+typedef rmw_ret_t (*rmw_send_request_t)(const rmw_client_t *client,
+                                        const void *ros_request,
+                                        int64_t *sequence_id);
+
+typedef rmw_ret_t (*rmw_take_request_t)(const rmw_service_t *service,
+                                        rmw_service_info_t *request_header,
+                                        void *ros_request, bool *taken);
+
+typedef rmw_ret_t (*rmw_send_response_t)(const rmw_service_t *service,
+                                         rmw_request_id_t *request_header,
+                                         void *ros_response);
+
+typedef rmw_ret_t (*rmw_take_response_t)(const rmw_client_t *client,
+                                         rmw_service_info_t *request_header,
+                                         void *ros_response, bool *taken);
+
+/************************************************************/
+/* Pointers to underlying versions of the hooked functions, */
+/* initialized on demand.                                   */
+/************************************************************/
+
+rmw_create_node_t real_rmw_create_node;
```

Comment:
These globals can be `static`, so they're only visible in this file that they're declared in